File: Two_Pointer/1537. Get the Maximum Score.cpp
/*
1537. Get the Maximum Score
*/

// TC:O(m+n) SC:O(1)
class Solution {
    using LL = long long;
    LL M = 1e9+7;
public:
    int maxSum(vector<int>& nums1, vector<int>& nums2) {
        LL sum1 = 0, sum2 = 0;
        int i = 0; int j = 0;
        while( i < nums1.size() && j < nums2.size()){
            if(nums1[i] < nums2[j]){
                sum1= sum1+nums1[i];
                i++;
            }
            else if(nums1[i] > nums2[j]){
                sum2= sum2+nums2[j];
                j++;
            }
            else if(nums1[i] == nums2[j]){
                sum1 = max(sum1,sum2)+nums1[i];
                sum2 = sum1;
                i++; j++;
            }
        }
        while(i < nums1.size()){
            sum1= sum1+nums1[i];
            i++;
        }
        while(j < nums2.size()){
            sum2= sum2+nums2[j];
            j++;
        }

        return (max(sum1,sum2))%M;
        
    }
};

// 補一個前綴和 + 分組取最大
class Solution {
    using LL = long long;
    LL MOD = 1e9 + 7;
public:
    int maxSum(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size(), m = nums2.size(), i = 0, j = 0;
        vector<LL> presum1(n+1), presum2(m+1);
        for(int i = 0; i < n; i++) presum1[i+1] = presum1[i] + (LL)nums1[i];
        for(int i = 0; i < m; i++) presum2[i+1] = presum2[i] + (LL)nums2[i];
        vector<pair<int,int>> arr{{0,0}};
        while(i < n && j < m) {
            if(nums1[i]<nums2[j]) i++;
            else if(nums1[i]>nums2[j]) j++;
            else {
                arr.emplace_back(i+1,j+1);
                i ++, j++;
            }
        }
        arr.emplace_back(n,m);
        LL ans = 0;
        for(int i = 1; i < arr.size();i++) {
            ans += max(presum1[arr[i].first] - presum1[arr[i-1].first], presum2[arr[i].second] - presum2[arr[i-1].second]);
        }
        return ans % MOD;
    }
};

File: Dynamic_Programming/1537. Get the Maximum Score.py
class Solution:
    def maxSum(self, nums1: List[int], nums2: List[int]) -> int:
        m, n = len(nums1), len(nums2)
        MOD = 10 ** 9 + 7
        # 空間優化
        f0, f1 = 0, 0
        i, j = 0, 0
        while i < m and j < n:
            if nums1[i] < nums2[j]:
                f0 += nums1[i]
                i += 1
            elif nums1[i] > nums2[j]:
                f1 += nums2[j]
                j += 1
            else:
                f1,f0 = max(f1,f0) + nums1[i], max(f1,f0) + nums1[i]
                i, j = i + 1, j + 1
        while i < m:
            f0 += nums1[i]
            i += 1
        while j < n:
            f1 += nums2[j]
            j += 1
        return max(f0, f1) % MO
        '''
        貪心雙指針 
        f0, f1 = [0] * (m+1), [0] * (n+1)
        i, j = 0, 0
        while i < m and j < n:
            if nums1[i] < nums2[j]:
                f0[i+1] = f0[i] + nums1[i]
                i += 1
            elif nums1[i] > nums2[j]:
                f1[j+1] = f1[j] + nums2[j]
                j += 1
            else:
                f1[j+1] = max(f1[j] + nums2[j], f0[i] + nums2[j])
                f0[i+1] = max(f0[i] + nums1[i], f1[j] + nums1[i])
                i += 1
                j += 1
        while i < m:
            f0[i+1] = f0[i] + nums1[i]
            i += 1
        while j < n:
            f1[j+1] = f1[j] + nums2[j]
            j += 1
        return max(f0[m], f1[n]) % MOD
        '''
        '''
        遞歸
        mp1 = {num: idx for idx, num in enumerate(nums1)}
        mp2 = {num: idx for idx, num in enumerate(nums2)}
        @cache
        def dfs(i:int, j:int) -> int:
            if (j == 0 and i == m) or (j == 1 and i == n):
                return 0
            ans = 0
            if j == 0:
                ans = max(dfs(mp2[nums1[i]] + 1, 1), dfs(i+1, 0)) + nums1[i] if nums1[i] in mp2 else dfs(i+1, 0) + nums1[i]
            else:
                ans = max(dfs(mp1[nums2[i]] + 1, 0), dfs(i+1, 1)) + nums2[i] if nums2[i] in mp1 else dfs(i+1, 1) + nums2[i]
            return ans
        return max(dfs(0, 0), dfs(0, 1)) % MOD
        '''
